vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,138 @@
+
+from vllm.config import VllmConfig
+from vllm.transformers_utils.tokenizer_group import init_tokenizer_from_configs
+from vllm.v1.request import RequestStatus
+
+if TYPE_CHECKING:
+    from vllm.config import ModelConfig
+    from vllm.transformers_utils.tokenizer_group import BaseTokenizerGroup
+    from vllm.v1.request import GuidedDecodingKey, Request
+
+
+class Grammar:
+
+    def __init__(self, matcher: xgr.GrammarMatcher, vocab_size: int,
+                 ctx: xgr.CompiledGrammar) -> None:
+        # https://xgrammar.mlc.ai/docs/api/python/index.html#xgrammar.GrammarMatcher.find_jump_forward_string for jump-forward decoding
+        # TODO: support max_rollback_tokens
+        self.matcher = matcher
+        self.vocab_size = vocab_size
+        self.ctx = ctx
+        self.prefilled = False
+
+    def accept_token(self, token: int) -> bool:
+        # NOTE: accept_token will determines whether we accept this token
+        # and will also update the machine state
+        return self.matcher.accept_token(token)
+
+    def allocate_bitmask(self, batch_size: int,
+                         vocab_size: int) -> torch.Tensor:
+        return xgr.allocate_token_bitmask(batch_size, vocab_size)
+
+    # this should be ran in parallel with model decoding
+    def fill_bitmask(self, bitmask: torch.Tensor, idx: int) -> None:
+        self.matcher.fill_next_token_bitmask(bitmask, idx)
+
+    @staticmethod
+    def apply_bitmask(logits: torch.Tensor, vocab_mask: torch.Tensor) -> None:
+        xgr.apply_token_bitmask_inplace(logits, vocab_mask)
+
+    def reset(self):
+        self.matcher.reset()
+
+    def copy(self):
+        return Grammar(matcher=xgr.GrammarMatcher(self.ctx),
+                       vocab_size=self.vocab_size,
+                       ctx=self.ctx)
+
+    def __copy__(self):
+        return self.copy()
+
+
+@dataclass
+class GrammarCache:
+    value: Optional[Grammar]
+    event: threading.Event
+
+
+class GuidedDecodingManager:
+
+    def __init__(self, vllm_config: VllmConfig):
+        tokenizer_group = init_tokenizer_from_configs(
+            model_config=vllm_config.model_config,
+            scheduler_config=vllm_config.scheduler_config,
+            parallel_config=vllm_config.parallel_config,
+            lora_config=vllm_config.lora_config)
+        tokenizer_group.ping()
+        self.model_config = vllm_config.model_config
+        self.vocab_size = vllm_config.model_config.get_vocab_size()
+        self.tokenizer = tokenizer_group.get_lora_tokenizer(None)
+        self.grammar_cache: dict[GuidedDecodingKey, GrammarCache] = {}
+        self.executor = ThreadPoolExecutor()
+        self._lock = threading.Lock()
+
+    def initialize_cache(self, key: GuidedDecodingKey) -> Grammar:
+        request_type, grammar_spec = key
+        tokenizer_info = xgr.TokenizerInfo.from_huggingface(
+            self.tokenizer, vocab_size=self.vocab_size)
+        compiler = xgr.GrammarCompiler(tokenizer_info, max_threads=8)
```

Comment:
The GrammarCompiler should be made once, stored in the initializer, and reused for all requests coming in.

Per the xgrammar definition, it has a built-in cache so we can remove the concept of GrammarCache here and simplify
https://github.com/mlc-ai/xgrammar/blob/02dff84aef7ac4192473a8f995a5a2938829c28a/python/xgrammar/compiler.py#L36-L40
```python
class GrammarCompiler(XGRObject):
    """The compiler for grammars. It is associated with a certain tokenizer info, and compiles
    grammars into CompiledGrammar with the tokenizer info. It allows parallel compilation with
    multiple threads, and has a cache to store the compilation result, avoiding compiling the
    same grammar multiple times.
    ...
    """
```